pkg/app/ops/deploymentchaincontroller/controller.go

```diff
@@ -0,0 +1,154 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package deploymentchaincontroller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipe/pkg/datastore"
+	"github.com/pipe-cd/pipe/pkg/model"
+)
+
+const (
+	// interval represents time to sync state for all deployment chain.
+	interval = time.Minute
+	// updaterWorkerNum represents the maximum number of updaters which can be
+	// run at the same time.
+	maxUpdaterWorkerNum = 10
+)
+
+type DeploymentChainController struct {
+	applicationStore     datastore.ApplicationStore
+	deploymentStore      datastore.DeploymentStore
+	deploymentChainStore datastore.DeploymentChainStore
+	// Map from deployment chain ID to the updater
+	// who in charge for the deployment chain updating.
+	updaters map[string]*updater
+
+	logger *zap.Logger
+}
+
+func NewDeploymentChainController(
+	ds datastore.DataStore,
+	logger *zap.Logger,
+) *DeploymentChainController {
+	return &DeploymentChainController{
+		applicationStore:     datastore.NewApplicationStore(ds),
+		deploymentStore:      datastore.NewDeploymentStore(ds),
+		deploymentChainStore: datastore.NewDeploymentChainStore(ds),
+		updaters:             make(map[string]*updater),
+		logger:               logger.Named("deployment-chain-controller"),
+	}
+}
+
+func (d *DeploymentChainController) Run(ctx context.Context) error {
+	t := time.NewTicker(interval)
+	defer t.Stop()
+
+	d.logger.Info("start running deployment chain controller")
+	defer d.logger.Info("deployment chain controller has been stopped")
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-t.C:
+			d.syncUpdaters(ctx)
+		}
+	}
+}
+
+func (d *DeploymentChainController) syncUpdaters(ctx context.Context) error {
+	// Remove done updater of completed deployment chain.
+	for id, u := range d.updaters {
+		if u.IsDone() {
+			d.logger.Info("remove done updater of deployment chain",
+				zap.String("id", id),
+				zap.Time("completed_at", u.doneTimestamp),
+			)
+			delete(d.updaters, id)
+		}
+	}
+
+	// Find all not completed deployment chains and create updater if does not exist.
+	notCompletedChains, err := listNotCompletedDeploymentChain(ctx, d.deploymentChainStore)
+	if err != nil {
+		d.logger.Error("failed to fetch all not completed deployment chain", zap.Error(err))
+		return err
+	}
+	for _, c := range notCompletedChains {
+		if _, ok := d.updaters[c.Id]; !ok {
+			d.updaters[c.Id] = newUpdater(
+				c,
+				d.applicationStore,
+				d.deploymentStore,
+				d.deploymentChainStore,
+				d.logger,
+			)
+		}
+	}
+
+	var (
+		dcUpdatersCtn = len(d.updaters)
```

Comment:
What does `Ctn` stand for?
nit: `updaterNum`

Reply:
I usually use `Ctn` as a counting variable, but do not have a strong opinion on this. Will rename it 👍